File: guests/go/richlabel/richlabel.go
// The rich label scene (tools/scenes/richlabel.steps; docs/rich-text-plan.md
// R8, §15): a label carries the inline vocabulary read-only — the app writes
// its document and edits it, and the widget draws the runs over the role's
// own font. THE OFFSETS ARE UTF-8 BYTES, which is what Go's own string
// indexing already produces (docs/ranges-units.md).
package richlabel

import (
	"fmt"
	"strings"

	kaya "dev.kaya/bindings/go"
)

const document = "Héllo world, code"

// spell is the core's spelling of runs (`expect_runs`), so the binding's
// document and the core's mirror are compared as one string.
func spell(runs []kaya.TextRun) string {
	parts := make([]string, 0, len(runs))
	for _, run := range runs {
		if run.IsFlag() {
			parts = append(parts, fmt.Sprintf("%d:%d %s", run.Range.Start, run.Range.End, run.Name))
			continue
		}
		parts = append(parts, fmt.Sprintf("%d:%d %s=%s", run.Range.Start, run.Range.End, run.Name, run.Value))
	}
	return strings.Join(parts, "|")
}

func App() *kaya.App {
	app := kaya.NewApp()

	var runs kaya.Signal[string]
	var body, heading kaya.Widget

	app.Build(func(tx *kaya.Tx) {
		tx.Window(0).Title("richlabel")
		runs = tx.Signal("")

		tx.Mount(tx.Column(func() {
			bodyText := tx.Signal("")
			headingText := tx.Signal("Heading with italic")

			body = tx.Label(bodyText).Rich().A11yID("body") // label#0
			heading = tx.Label(headingText).Role(kaya.RoleHeading).
				Rich().A11yID("heading") // label#1
			tx.Label(runs).A11yID("runs") // label#2

			tx.Row(func() {
				tx.Button("seed", func(tx *kaya.Tx) { // button#0
					doc := kaya.NewDocument(document).
						Bold(0, 6).
						Link(7, 12, "https://kaya.dev").
						Flag(14, 18, "code", true)
					title := kaya.NewDocument("Heading with italic").
						Flag(13, 19, "italic", true)
					tx.SetDocument(body, doc)
					tx.SetDocument(heading, title)
					tx.Write(runs, spell(doc.Runs))
				})
				tx.Button("insert", func(tx *kaya.Tx) { // button#1
					tx.ApplyEdit(body, kaya.Insert(6, ", big").Flag(2, 5, "italic", true))
					tx.Write(runs, spell(app.Document(body).Runs))
				})
				tx.Button("mark", func(tx *kaya.Tx) { // button#2
					tx.FormatRangeFlag(body, kaya.TextRange{Start: 1, End: 4}, "italic", true)
					tx.UnformatRange(body, kaya.TextRange{Start: 0, End: 3}, "bold") // "Hé": byte 2 is inside the é
					tx.Write(runs, spell(app.Document(body).Runs))
				})
			})
		}))
	})

	return app
}

File: guests/java/dev/kaya/guests/Richlabel.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

import java.util.ArrayList;
import java.util.List;

/**
 * The rich label scene from the JVM — guests/rust/richlabel.rs,
 * tools/scenes/richlabel.steps (docs/rich-text-plan.md R8, §15): a label
 * carries the inline vocabulary read-only. THE OFFSETS ARE UTF-8 BYTES;
 * the accented first word is what makes a UTF-16 reader fail
 * (docs/ranges-units.md), so every range here goes through TextRange.in,
 * which converts against the text it indexes.
 */
public final class Richlabel {
    /** Java lambdas cannot assign captured locals. */
    private static final class Refs {
        KayaApp.Widget body, heading;
    }

    /** THE ACCENT IS IN A {@code \}{@code u} ESCAPE: a literal's bytes
     * follow the compiler's encoding, and the Android build that also
     * compiles this directory passes no {@code -encoding}
     * (android/javahost/build.gradle.kts). */
    private static final String DOC = "H\u00e9llo world, code";

    private static final String TITLE = "Heading with italic";

    private static final String URL = "https://kaya.dev";

    /** The core's spelling of runs ({@code expect_runs}), so the
     * binding's document and the core's mirror are compared as one
     * string. */
    private static String spell(List<KayaApp.TextRun> runs) {
        List<String> parts = new ArrayList<>();
        for (KayaApp.TextRun run : runs) {
            parts.add(run.value().equals("true")
                    ? run.start() + ":" + run.stop() + " " + run.name()
                    : run.start() + ":" + run.stop() + " " + run.name() + "=" + run.value());
        }
        return String.join("|", parts);
    }

    public static void app() {
        KayaApp app = new KayaApp();

        app.build(tx -> {
            tx.window(0).title("richlabel");
            KayaApp.Signal<String> runs = tx.signal("");

            Refs refs = new Refs();
            tx.mount(tx.column(() -> {
                KayaApp.Signal<String> bodyText = tx.signal("");
                KayaApp.Signal<String> headingText = tx.signal(TITLE);

                refs.body = tx.label(bodyText).rich().a11yId("body"); // label#0
                refs.heading = tx.label(headingText).role(KayaApp.Role.HEADING)
                        .rich().a11yId("heading"); // label#1
                tx.label(runs).a11yId("runs"); // label#2

                tx.row(() -> {
                    tx.button("seed", t -> { // button#0
                        KayaApp.Document doc = new KayaApp.Document(DOC)
                                .bold(KayaApp.TextRange.in(DOC, 0, 5))
                                .link(KayaApp.TextRange.in(DOC, 6, 11), URL)
                                .mark(KayaApp.TextRange.in(DOC, 13, 17), "code", "true");
                        KayaApp.Document title = new KayaApp.Document(TITLE)
                                .mark(KayaApp.TextRange.in(TITLE, 13, 19), "italic",
                                        "true");
                        t.setDocument(refs.body, doc);
                        t.setDocument(refs.heading, title);
                        t.write(runs, spell(doc.runs()));
                    });
                    tx.button("insert", t -> { // button#1
                        KayaApp.Edit edit =
                                KayaApp.Edit.insert(KayaApp.TextRange.in(DOC, 5, 5), ", big")
                                        .mark(KayaApp.TextRange.in(", big", 2, 5), "italic",
                                                "true");
                        t.applyEdit(refs.body, edit);
                        t.write(runs, spell(app.document(refs.body).runs()));
                    });
                    // THE RANGED ACT ON A LABEL (docs/rich-text-plan.md §17):
                    // the label's own document written by range, no selection
                    // to move; the ranges convert against the CURRENT text,
                    // which the insert moved.
                    tx.button("mark", t -> { // button#2
                        String text = app.document(refs.body).text();
                        t.formatRange(refs.body, KayaApp.TextRange.in(text, 1, 3), "italic",
                                "true");
                        t.unformatRange(refs.body, KayaApp.TextRange.in(text, 0, 2), "bold");
                        t.write(runs, spell(app.document(refs.body).runs()));
                    });
                });
            }));
        });

        app.dispatchLoop();
    }

    private Richlabel() {}
}
